File: backend/src/main/java/com/autocare/client/service/ClientService.java
package com.autocare.client.service;

import com.autocare.client.dto.ClientRequestDTO;
import com.autocare.client.dto.ClientResponseDTO;
import com.autocare.client.entity.Client;
import com.autocare.client.repository.ClientRepository;
import com.autocare.shared.exception.ResourceNotFoundException;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class ClientService {

    private final ClientRepository clientRepository;

    @Transactional
    public ClientResponseDTO create(ClientRequestDTO request) {
        Client client = Client.builder()
                .name(request.getName())
                .cpf(request.getCpf())
                .phone(request.getPhone())
                .email(request.getEmail())
                .address(request.getAddress())
                .isActive(true)
                .build();

        client = clientRepository.save(client);
        return toResponseDTO(client);
    }

    @Transactional
    public ClientResponseDTO update(String id, ClientRequestDTO request) {
        Client client = findById(id);

        client.setName(request.getName());
        client.setCpf(request.getCpf());
        client.setPhone(request.getPhone());
        client.setEmail(request.getEmail());
        client.setAddress(request.getAddress());

        client = clientRepository.save(client);
        return toResponseDTO(client);
    }

    @Transactional
    public void delete(String id) {
        Client client = findById(id);
        client.setActive(false);
        clientRepository.save(client);
    }

    @Transactional
    public void activate(String id) {
        Client client = findById(id);
        client.setActive(true);
        clientRepository.save(client);
    }

    public ClientResponseDTO findByIdResponse(String id) {
        return toResponseDTO(findById(id));
    }

    public List<ClientResponseDTO> findAll() {
        return clientRepository.findByIsActiveTrue()
                .stream()
                .map(this::toResponseDTO)
                .collect(Collectors.toList());
    }

    public List<ClientResponseDTO> searchByName(String name) {
        return clientRepository.findByNameContainingIgnoreCase(name)
                .stream()
                .map(this::toResponseDTO)
                .collect(Collectors.toList());
    }

    public Client findById(String id) {
        return clientRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Cliente não encontrado com ID: " + id));
    }

    private ClientResponseDTO toResponseDTO(Client client) {
        return ClientResponseDTO.builder()
                .id(client.getId())
                .name(client.getName())
                .cpf(client.getCpf())
                .phone(client.getPhone())
                .email(client.getEmail())
                .address(client.getAddress())
                .createdAt(client.getCreatedAt())
                .isActive(client.isActive())
                .vehicleCount(client.getVehicles() != null ? client.getVehicles().size() : 0)
                .build();
    }
}

File: backend/src/client/client.service.ts
import { Injectable } from "@nestjs/common";
import { PrismaService } from "../prisma/prisma.service";
import { ResourceNotFoundException } from "../common/exceptions";
import { ClientRequestDto } from "./dto/client.dto";

@Injectable()
export class ClientService {
  constructor(private prisma: PrismaService) {}

  async create(dto: ClientRequestDto) {
    const client = await this.prisma.client.create({
      data: { ...dto, isActive: true },
      include: { vehicles: true },
    });
    return this.toResponse(client);
  }

  async update(id: string, dto: ClientRequestDto) {
    await this.findById(id);
    const client = await this.prisma.client.update({
      where: { id },
      data: dto,
      include: { vehicles: true },
    });
    return this.toResponse(client);
  }

  async delete(id: string) {
    await this.findById(id);
    await this.prisma.client.update({ where: { id }, data: { isActive: false } });
  }

  async activate(id: string) {
    await this.findById(id);
    await this.prisma.client.update({ where: { id }, data: { isActive: true } });
  }

  async findByIdResponse(id: string) {
    return this.toResponse(await this.findByIdWithVehicles(id));
  }

  async findAll() {
    const clients = await this.prisma.client.findMany({
      where: { isActive: true },
      include: { vehicles: true },
      orderBy: { createdAt: "desc" },
    });
    return clients.map((c) => this.toResponse(c));
  }

  async searchByName(name: string) {
    const clients = await this.prisma.client.findMany({
      where: { name: { contains: name, mode: "insensitive" } },
      include: { vehicles: true },
    });
    return clients.map((c) => this.toResponse(c));
  }

  // Equivalente a ClientService.findById do backend original: retorna a entidade
  // crua (não o DTO de resposta), usado internamente por outros services
  // (VehicleService, ServiceOrderService, BudgetService).
  async findById(id: string) {
    const client = await this.prisma.client.findUnique({ where: { id } });
    if (!client) {
      throw new ResourceNotFoundException(`Cliente não encontrado com ID: ${id}`);
    }
    return client;
  }

  private async findByIdWithVehicles(id: string) {
    const client = await this.prisma.client.findUnique({
      where: { id },
      include: { vehicles: true },
    });
    if (!client) {
      throw new ResourceNotFoundException(`Cliente não encontrado com ID: ${id}`);
    }
    return client;
  }

  private toResponse(client: any) {
    return {
      id: client.id,
      name: client.name,
      cpf: client.cpf,
      phone: client.phone,
      email: client.email,
      address: client.address,
      createdAt: client.createdAt,
      isActive: client.isActive,
      vehicleCount: client.vehicles ? client.vehicles.length : 0,
    };
  }
}
